lib/puppet/util/windows/eventlog.rb

```diff
@@ -0,0 +1,180 @@
+    end
+  end
+
+  # Close this instance's event log handle
+  # @return [void]
+  # @api public
+  def close
+    result_of_close = DeregisterEventSource(@eventlog_handle)
+    if result_of_close == WIN32_FALSE
+      raise EventlogError.new("failed to close Windows eventlog")
+    end
+    @eventlog_handle = nil
+  end
+
+  # Report an event to this instance's event log handle. Accepts a string to
+  #   report (:data => <string>) and event type (:event_type => FixNum) and id
+  # (:event_id => FixNum) as returned by #to_native. The additional arguments to
+  # ReportEventW seen in this method aren't exposed - though ReportEventW
+  # technically can accept multiple strings as well as raw binary data to log,
+  # we accept a single string from Puppet::Util::Log
+  #
+  # @param args [Hash{Symbol=>Object}] options to the associated log event
+  # @return [void]
+  # @api public
+  def report_event(args = {})
+    raise ArgumentError, "data must be a string, not #{args[:data].class}" unless args[:data].is_a?(String)
+    from_string_to_wide_string(args[:data]) do |message_ptr|
+      FFI::MemoryPointer.new(:pointer, 2) do |message|
+        message[0].write_pointer(message_ptr)
+        user_sid, raw_data = nil
+        raw_data_size = 0
+        num_strings = 1
+        eventlog_category = 0
+        report_result = ReportEventW(@eventlog_handle, args[:event_type],
+          eventlog_category, args[:event_id], user_sid,
+          num_strings, raw_data_size, message, raw_data)
+
+        if report_result == WIN32_FALSE
+          raise EventlogError.new("failed to report event to Windows eventlog")
+        end
+      end
+    end
+  end
+
+  class << self
+    # Feels more natural to do Puppet::Util::Window::EventLog.open("MyApplication")
+    alias :open :new
+
+    # Query event identifier info for a given log level
+    # @param level [Symbol] an event log level
+    # @return [Array] Win API Event ID, Puppet Event ID
+    # @api public
+    def to_native(level)
+      case level
+      when :debug,:info,:notice
+        [EVENTLOG_INFORMATION_TYPE, 0x01]
+      when :warning
+        [EVENTLOG_WARNING_TYPE, 0x02]
+      when :err,:alert,:emerg,:crit
+        [EVENTLOG_ERROR_TYPE, 0x03]
+      else
+        raise ArgumentError, "Invalid log level #{level}"
+      end
+    end
+  end
+
+  private
+  # For the purposes of allowing this class to be standalone, the following are
+  # duplicate definitions from elsewhere in Puppet:
+  NULL_HANDLE = 0
+  WIN32_FALSE = 0
+
+  # If we're loaded via Puppet we should keep the previous behavior of raising
+  # Puppet::Util::Windows::Error on errors. For daemon.rb we don't have
+  # Puppet::Util::Windows::Error, but daemon.rb just silently blanket rescues
+  # all Exceptions raised during a log attempt, so we're fine just raising a
+  # generic RuntimeError.
+  EventlogError = defined?(Puppet::Util::Windows::Error) ? Puppet::Util::Windows::Error : RuntimeError
```

Comment:
Just to make sure I understand, the concern about not capturing `FFI.errno` / `FormatMessage` would only apply to situations where `Puppet::Util::Windows::Error` is not defined/available (since we'll use it if it is)? If that's the case, I think I'm fine with leaving it a generic runtime error when `Puppet::Util::Windows::Error` isn't available. The reason being, the only place (at this time) outside of Puppet where we use this code is in `ext/windows/services/daemon.rb`, for running the puppet service itself, and `daemon.rb` just entirely ignores any failure to log something: https://github.com/puppetlabs/puppet/blob/master/ext/windows/service/daemon.rb#L151-L153. If we build on the use cases in the future I'd be +1 to extending the error handling.